File: frontend/src/app/core/services/category.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { ApiResponse } from '../models/user.model';
import { CategoryRequest, CategoryResponse } from '../models/category.model';

@Injectable({
  providedIn: 'root',
})
export class CategoryService {
  private readonly http = inject(HttpClient);

  getCategoryById(categoryId: string): Observable<ApiResponse<CategoryResponse>> {
    return this.http.get<ApiResponse<CategoryResponse>>(`/api/v1/categories/${categoryId}`);
  }

  createCategory(request: CategoryRequest): Observable<ApiResponse<CategoryResponse>> {
    return this.http.post<ApiResponse<CategoryResponse>>('/api/v1/categories', request);
  }

  updateCategory(categoryId: string, request: CategoryRequest): Observable<ApiResponse<CategoryResponse>> {
    return this.http.put<ApiResponse<CategoryResponse>>(`/api/v1/categories/${categoryId}`, request);
  }

  deleteCategory(categoryId: string): Observable<ApiResponse<void>> {
    return this.http.delete<ApiResponse<void>>(`/api/v1/categories/${categoryId}`);
  }
}

File: backend/src/main/java/com/chermew/todolist/service/CategoryService.java
package com.chermew.todolist.service;

import com.chermew.todolist.dto.*;

import java.util.List;
import java.util.UUID;

public interface CategoryService {
    CategoryResponse createCategory(CategoryRequest request, UUID userId);
    CategoryResponse updateCategory(UUID categoryId, CategoryRequest request, UUID userId);
    CategoryResponse getCategoryById(UUID categoryId, UUID userId);
    void deleteCategory(UUID categoryId, UUID id);
}
